samples/bluetooth/hci_uart_async/src/hci_uart_async.c

```diff
@@ -0,0 +1,371 @@
+/* Copyright (c) 2023 Nordic Semiconductor ASA
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <zephyr/kernel.h>
+#include <zephyr/sys/__assert.h>
+#include <zephyr/sys/time_units.h>
+#include <zephyr/toolchain/common.h>
+#include <zephyr/drivers/bluetooth/hci_driver.h>
+#include <errno.h>
+#include <stddef.h>
+#include <string.h>
+
+#include <zephyr/kernel.h>
+#include <zephyr/arch/cpu.h>
+#include <zephyr/sys/byteorder.h>
+#include <zephyr/logging/log.h>
+#include <zephyr/sys/util.h>
+
+#include <zephyr/device.h>
+#include <zephyr/init.h>
+#include <zephyr/drivers/uart.h>
+
+#include <zephyr/usb/usb_device.h>
+
+#include <zephyr/net/buf.h>
+#include <zephyr/bluetooth/bluetooth.h>
+#include <zephyr/bluetooth/l2cap.h>
+#include <zephyr/bluetooth/hci.h>
+#include <zephyr/bluetooth/buf.h>
+#include <zephyr/bluetooth/hci_raw.h>
+
+LOG_MODULE_REGISTER(hci_uart_async, LOG_LEVEL_DBG);
+
+static const struct device *const hci_uart_dev = DEVICE_DT_GET(DT_CHOSEN(zephyr_bt_c2h_uart));
+
+static K_THREAD_STACK_DEFINE(h2c_thread_stack, CONFIG_BT_HCI_TX_STACK_SIZE);
+static struct k_thread h2c_thread;
+
+enum h4_type {
+	H4_CMD = 0x01,
+	H4_ACL = 0x02,
+	H4_SCO = 0x03,
+	H4_EVT = 0x04,
+	H4_ISO = 0x05,
+};
+
+struct k_poll_signal uart_h2c_rx_sig;
+struct k_poll_signal uart_c2h_tx_sig;
+
+static K_FIFO_DEFINE(c2h_queue);
+
+static int uart_c2h_tx(const uint8_t *data, size_t size)
+{
+	int err;
+
+	struct k_poll_signal *sig = &uart_c2h_tx_sig;
+
+	struct k_poll_event done[] = {
+		K_POLL_EVENT_INITIALIZER(K_POLL_TYPE_SIGNAL, K_POLL_MODE_NOTIFY_ONLY, sig),
+	};
+
+	/* Maybe take a mutex? */
```

Comment:
would be good to answer that question.

Reply:
I marked this function not-thread-safe. It's only invoked from one thread, so there is no need for a mutex.